fuel-relayer/src/relayer.rs

```diff
@@ -154,55 +141,24 @@ impl Relayer {
         );
 
         for start in (last_finalized_da_height..best_finalized_block).step_by(step) {
-            let end = min(start + step as u64, best_finalized_block);
-            if (start - last_finalized_da_height) % REPORT_PROGRESS_EVERY_N_BLOCKS == 0 {
-                info!("geting log from:{}", start);
+            let end = min(start + step as DaBlockHeight, best_finalized_block);
+            if (start - last_finalized_da_height) % config::REPORT_INIT_SYNC_PROGRESS_EVERY_N_BLOCKS
+                == 0
+            {
+                info!("geting log from height:{}", start);
             }
 
             // TODO  can be parallelized
             let filter = Filter::new()
                 .from_block(start)
                 .to_block(end)
                 .address(ValueOrArray::Array(contracts.clone()));
-            let logs = self.stop_handle(|| provider.get_logs(&filter)).await??;
-
-            for eth_event in logs {
-                let fuel_event = EthEventLog::try_from(&eth_event);
-                if let Err(_err) = fuel_event {
-                    // not formated event from contract
-                    // just skip it for now.
-                    continue;
-                }
-                let fuel_event = fuel_event.unwrap();
-                self.pending
-                    .append_eth_events(&fuel_event, eth_event.block_number.unwrap().as_u64())
-                    .await;
-            }
-            // if there is more then two items in pending list flush first one.
-            // Having two elements in this stage means that full fuel block is already processed and
-            // we dont have reverts to dispute that.
-            while self.pending.len() > 1 {
-                // we are sending dummy eth block bcs we are sure that it is finalized
-                self.pending
-                    .commit_diffs(self.db.as_mut(), best_finalized_block)
-                    .await;
-            }
-        }
+            let logs = handle_interrupt!(self, provider.get_logs(&filter))??;
+            self.queue.append_eth_logs(logs).await;
 
-        // if there is no diffs it means we are at start of contract creating
-        let last_diff = if self.pending.is_empty() {
-            // set fuel num to zero and contract creating eth.
-            PendingDiff::new(0)
-        } else {
-            // apply all pending changed.
-            while self.pending.len() > 1 {
-                // we are sending dummy eth block num bcs we are sure that it is finalized
-                self.pending
-                    .commit_diffs(self.db.as_mut(), best_finalized_block)
-                    .await;
-            }
-            self.pending.pop_back().unwrap()
-        };
+            // we are sending dummy eth block bcs we are sure that it is finalized
+            self.queue.commit_diffs(self.db.as_mut(), end).await;
+        }
```

Comment:
I have an idea to simplify this process and make the relayer a bit more durable.

If I understand this next section of code properly, we're trying to backfill all the unfinalized events so that we'll be ready to just start streaming/appending new logs as they're observed. We have these retries in an attempt to ensure:
1. No logs are skipped between initialization and active log listening during regular operation
2. Protect against potential re-orgs.

What if we instead go with the following approach:

Whenever a new log is received, we perform the necessary checks to ensure there were no gaps between the last events we received (e.g. check that block number delta == 1, or that parent block hash matches our last block hash etc), and if we detect any jumps, we fetch the missing events using `get_logs()` until we're caught up to the event we just received.

In the happy path, get_logs is never called, most of these lookups could be done using cached data, and then we simply just append the log.

In the unhappy path, we stop polling for new events until we are resync'd.

Initialization could be drastically simplified to simply grabbing and storing all the finalized events, and then we consolidate concerns like handling pending events and reorgs to a single location. This would also be more reliable in the event some messages were dropped between the eth client subscription and the relayer, as the relayer would be able to automatically detect a divergence in expected state and retrieve the missing history.

Reply:
Another more radical idea is to ignore any events beyond the finalization window simply. We could just subscribe only to new block events and then call `get_logs()` on (new block - finalization period). Maybe not as efficient as streaming each event via pubsub, but we wouldn't need all the complicated queueing logic to handle reorgs or ever worry about dropped messages.

Reply:
Having pending logs (except for block commit) is to have additional information on pending things, it is leftover from the first iteration and it is a working mechanism.
The second proposal can act as an addon to what we already have. As you said we can check finalized blocks and ask eth client for logs on that block this is a more robust way to do things, but I didn't implement it as current mechanism should work just fine, and moving my focus to build missing parts of the client seems like a higher priority.

Reply:
> moving my focus to build missing parts of the client seems like a higher priority.

I'm not saying building the rest of the client is low priority, but we should be making our best effort to ensure loose ends are tied up in one area before opening several other cans of worms. Otherwise, the cognitive overhead of tracking how "done" things actually are will be too hard to manage and we may end up shipping unstable products.